### PR TITLE
Prepare for selective UIA event registration by default

### PR DESCRIPTION
**Pull request by codeofdusk**
### Link to issue number:
None

### Summary of the issue:
We would like to switch to selective UIA event registration by default.

### Description of how this pull request fixes the issue:
Convert the selective UIA registration option to a ternary setting and factor out logic to determine whether to register selectively into a new function.

### Testing strategy:
N/A

### Known issues with pull request:
None known

### Change log entries:
Not user-facing

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English

### CODE REVIEW

**Comment by leonardder** [comment]
There is a known issue that task manager is misbehaving when selective registration is on. I guess we'll have to take that for ok or fix it somehow.

**Comment by seanbudd** [comment]
@leonardder - you are referring to #11599, correct?

I searched NVDA issues for "selective registration"

@codeofdusk - do you mind investigating how this PR effects #11599, #11354, #11109, #8624? 

It looks like this will improve the latter two, but make behaviour worse for the first two.

**Comment by josephsl** [comment]
Hi, note that enabling selective UIA event registration will break Windows 10 emoji panel support (does not affect Windows 11) due to the fact that emoji panel interface is an overlay that does not take system focus at all. Thanks.

**Comment by codeofdusk** [comment]
* #11599: Can still reproduce.
* #11354: cannot reproduce.
* #11109 and #8624: more specific instances of #11002 (which is improved).

**Comment by codeofdusk** [comment]
> Another path is to create a third state for the setting, "Windows 11 only" which only enables this on Windows 11.

@seanbudd This would require a config spec schema upgrade. If we can fix #11599 I think that could be reasonable.

**Comment by codeofdusk** [comment]
Actually, it looks like I can no longer reproduce #11599 at least on Nickel. I'll proceed with the config spec upgrade.

**Comment by leonardder** [comment]
Can we use the new feature flag approach while at it?

**Comment by codeofdusk** [comment]
> Can we use the new feature flag approach while at it?

I'm not sure. This seems closer to the "diff algorithm" or "Windows Console support" option (choice of two approaches or an "automatic" option that lets NVDA decide based on runtime environment) than an opt-in/opt-out feature with set NVDA developer default.

**Review by Qchristensen — APPROVED** [review]
User Guide change looks good

**Comment by codeofdusk** [comment]
Given the uncertainty around #11599 and the emoji panel, I'd like to limit this to SV2 as I know that everything works as expected.

Blocked by #13867.

File: source/UIAHandler/utils.py
Comment (lukaszgo1) [inline source/UIAHandler/utils.py:357]:
```suggestion
	"Determines whether to register for UIA events selectively or globally."
```

**Comment by codeofdusk** [comment]
#13867 has been merged, so this PR is no longer blocked.

CC @seanbudd.

**Comment by seanbudd** [comment]
@codeofdusk can you update the PR description? it appears out of date

**Comment by codeofdusk** [comment]
@seanbudd What do you mean? It looks up-to-date to me.

File: source/config/profileUpgradeSteps.py
Comment (seanbudd) [inline source/config/profileUpgradeSteps.py:104]: While inconsistent with other function definitions, I believe this typing is more accurate

```suggestion
def upgradeConfigFrom_6_to_7(profile: Dict[str, str]) -> None:
```

File: source/config/profileUpgradeSteps.py
Comment (seanbudd) [inline source/config/profileUpgradeSteps.py:107]: What does `profile['UIA']['eventRegistration']` become as the result of upgrading when:
- the user has not changed the default value of `profile['UIA']['selectiveEventRegistration']` (e.g. it is set to False)
- the user has explicitly set the value to False

I currently only understand the case where:
- the user has explicitly set the value to True

Could you add a summary of what happens for the 3 cases to the docstring?

**Comment by seanbudd** [comment]
@codeofdusk the summary of the issue reads as if the bugs are still an issue. Can you add that these bugs are fixed in Windows 11 SV2?

File: source/config/profileUpgradeSteps.py
Comment (seanbudd) [inline source/config/profileUpgradeSteps.py:104]: Could you also update the copyright headers for this file?

File: source/config/profileUpgradeSteps.py
Comment (codeofdusk) [inline source/config/profileUpgradeSteps.py:104]: For some reason, this change prevents NVDA from starting on my system.

**Comment by codeofdusk** [comment]
@seanbudd For some reason, adding your type hints to the config upgrade step seem to prevent NVDA from starting on my machine.

File: source/config/profileUpgradeSteps.py
Comment (seanbudd) [reply source/config/profileUpgradeSteps.py:104]: You need to import `Dict` from `typing`

File: source/config/profileUpgradeSteps.py
Comment (seanbudd) [inline source/config/profileUpgradeSteps.py:3]: To fix linting you will probably have to add a space after the `#` for each line

File: source/config/profileUpgradeSteps.py
Comment (seanbudd) [inline source/config/profileUpgradeSteps.py:18]: For typing imports, we tend to use the following syntax, as the import list tends to grow quickly and this makes diffs easier in future

```suggestion
from typing import (
	Dict,
)
```

**Review by seanbudd — APPROVED** [review]
Thanks @codeofdusk  LGTM once these minor things get fixed!

**Comment by codeofdusk** [comment]
@seanbudd Do you think it's alright to keep the config spec upgrade and behaviour change in one PR, and in the event of revert follow the process outlined in the known issues? Or should I separate them similar to UIA console?

**Comment by seanbudd** [comment]
@codeofdusk - that would be ideal, but I believe there is less risk of reverting for this PR

**Comment by codeofdusk** [comment]
@seanbudd Done. I'll make a follow-up once this is merged.

**Comment by codeofdusk** [comment]
Thanks @seanbudd. I've opened #14018.

**Pull request merged**
